x-pack/plugins/security/server/authentication/authenticator.ts

```diff
@@ -325,6 +325,9 @@ export class Authenticator {
           const auditLogger = this.options.audit.asScoped(request);
           auditLogger.log(
             userLoginEvent({
+              // We must explicitly specify the sessionId for login events because we just created the session, so
+              // it won't automatically get included in the audit event from the request context.
+              sessionId: sessionUpdateResult?.value?.sid,
```

Comment:
@azasypkin I haven't tested with SAML (I don't have it set up locally yet 🙈) but I am assuming this will work for intermediate sessions, since at this point we should have just updated the session to reflect that the user is fully authenticated. Is that correct? Or do I need to change it to something like this?

```ts
sessionId: sessionUpdateResult?.value?.sid || existingSessionValue?.sid,
```

Reply:
> since at this point we should have just updated the session to reflect that the user is fully authenticated. Is that correct?

When we turn unauthenticated (intermediate) session to an authenticated one, we basically [invalidate existing session](https://github.com/elastic/kibana/blob/0491214a57f2cf82e8b8e15734695ca3862f33b5/x-pack/plugins/security/server/authentication/authenticator.ts#L687-L691) and [create a new one](https://github.com/elastic/kibana/blob/0491214a57f2cf82e8b8e15734695ca3862f33b5/x-pack/plugins/security/server/authentication/authenticator.ts#L699-L704). 

So in case of successful login we'll have a new SID here `sessionUpdateResult?.value?.sid` which I think is what we want. But if we also want to record a SID of intermediate session if login fails, then we'd need this `sessionId: sessionUpdateResult?.value?.sid || existingSessionValue?.sid`. Is it something we want or not?

Reply:
> When we turn unauthenticated (intermediate) session to an authenticated one, we basically [invalidate existing session](https://github.com/elastic/kibana/blob/0491214a57f2cf82e8b8e15734695ca3862f33b5/x-pack/plugins/security/server/authentication/authenticator.ts#L687-L691) and [create a new one](https://github.com/elastic/kibana/blob/0491214a57f2cf82e8b8e15734695ca3862f33b5/x-pack/plugins/security/server/authentication/authenticator.ts#L699-L704).

Ah, thank you for clarifying!

> So in case of successful login we'll have a new SID here `sessionUpdateResult?.value?.sid` which I think is what we want.

Yes!

> But if we also want to record a SID of intermediate session if login fails, then we'd need this `sessionId: sessionUpdateResult?.value?.sid || existingSessionValue?.sid`. Is it something we want or not?

I don't think we need that, no. It won't show up in other audit records so it won't be useful for correlation.

Reply:
By the way, just tested SAML and I see that we don't record `user_login` for the intermediate sessions, but at the same time we record `user_logout` for them. Is it a big deal for the audit logs consumers?

Intermediate\unauthenticated session has been created:

```json
{
	"_index" : ".kibana_security_session_1",
	"_id" : "jp+qNp5O8J5EfuoaFqP79wI+u/reimJnXCc4eJktYUQ=",
	"_score" : 1.0,
	"_source" : {
	  "provider" : {
	    "type" : "saml",
	    "name" : "saml1"
	  },
	  "idleTimeoutExpiration" : 1643909890350,
	  "lifespanExpiration" : 1646473090350,
	  "content" : "...."
	}
}
```


When user is returned back to Kibana with SAML Response:

```json5
# Invalidate intermediate session
{"event":{"action":"user_logout","category":["authentication"],"outcome":"unknown"},"kibana":{"space_id":"default","session_id":"jp+qNp5O8J5EfuoaFqP79wI+u/reimJnXCc4eJktYUQ=","authentication_provider":"saml1","authentication_type":"saml"},"trace":{"id":"3b70b10e-76c3-4927-81b5-0ad1cd29f5cb"},"ecs":{"version":"8.0.0"},"@timestamp":"2022-02-03T10:39:41.248+01:00","message":"User [undefined] is logging out using saml provider [name=saml1]","log":{"level":"INFO","logger":"plugins.security.audit.ecs"},"process":{"pid":36133},"transaction":{"id":"827f8d12207b282d"}}


# Create a new authenticated session
{"event":{"action":"user_login","category":["authentication"],"outcome":"success"},"user":{"name":"saml-tester@xyz.abcdefg","roles":[]},"kibana":{"session_id":"tkJ4Vr7vn3Blfpg8AapjHzMrFxwViRsIevZ6/WM8GPg=","authentication_provider":"saml1","authentication_type":"saml","authentication_realm":"saml1","lookup_realm":"saml1"},"trace":{"id":"3b70b10e-76c3-4927-81b5-0ad1cd29f5cb"},"ecs":{"version":"8.0.0"},"@timestamp":"2022-02-03T10:39:43.125+01:00","message":"User [saml-tester@xyz.abcdefg] has logged in using saml provider [name=saml1]","log":{"level":"INFO","logger":"plugins.security.audit.ecs"},"process":{"pid":36133},"transaction":{"id":"827f8d12207b282d"}}
```

Reply:
> By the way, just tested SAML and I see that we don't record `user_login` for the intermediate sessions, but at the same time we record `user_logout` for them. Is it a big deal for the audit logs consumers?

Good catch, I'm glad you tested it!

I don't think we should be recording `user_logout` events when intermediate sessions are deleted. This really provides no value to an auditor, IMO.

Reply:
OK, in making this change I realized our tests weren't structured ideally --
The audit event assertions weren't fine-grained enough.
I added 066b67d2bf76ea300d0429c4de5b8bf5083154e3 first to restructure our tests and add more audit event assertions. Then I added 76c7d60e63fc435fd70b8b5cae2907da97a4bd23 to stop recording `user_logout` events for intermediate sessions that are deleted.